File: graph_solution.py
#!/usr/bin/env python3
"""
最長片道きっぷの旅の実装

可読性・リファクタリング性を上げるために、42Tokyoの各関数25行制限・引数3つまでのコーディング規則を採用しました。
引数の数を3つ以内に制限するために、初期実装の手続き型の実装をクラスベースにリファクタリング

"""

import sys
from collections import defaultdict
from typing import List, Tuple, Dict, Optional


# =======================================
# 入力処理（クラス外部の独立した関数）
# =======================================

def parse_line(line: str) -> Optional[Tuple[int, int, float]]:
    """1行をパースして(駅ID1, 駅ID2, 距離)を返す"""
    cleaned_line = line.strip()
    if not cleaned_line:
        return None
    try:
        parts = cleaned_line.split(',')
        if len(parts) != 3:
            return None
        u, v, weight = int(parts[0]), int(parts[1]), float(parts[2])
        return u, v, weight
    except (ValueError, IndexError):
        return None

def parse_input(lines: List[str]) -> List[Tuple[int, int, float]]:
    """標準入力からエッジリストをパースする"""
    return [edge for line in lines if (edge := parse_line(line)) is not None]


# =======================================
# メインのソルバークラス
# =======================================

class LongestPathSolver:
    """
    ビットマスクDPを用いてグラフの最長経路を計算するソルバー。
    関連する状態とロジックをすべてカプセル化する。
    """

    def __init__(self, edges: List[Tuple[int, int, float]]):
        # グラフ基本情報
        self.graph: Dict[int, Dict[int, float]] = defaultdict(dict)
        self.nodes: List[int] = []
        self.node_to_idx: Dict[int, int] = {}
        self._build_graph(edges)
        self.n = len(self.nodes)

        # DPテーブル
        self.dp: List[List[float]] = []
        self.parent: List[List[Optional[Tuple[int, int]]]] = []
        self.start_node: List[List[int]] = []
        self._initialize_dp()

        # 最長経路の情報
        self.max_distance: float = 0.0
        self.best_mask: int = 0
        self.best_end_idx: int = 0
        self.best_is_cycle: bool = False

    # ----------------------------------------
    # 初期化ヘルパー
    # ----------------------------------------

    def _build_graph(self, edges: List[Tuple[int, int, float]]) -> None:
        """無向グラフを構築し、ノード情報を設定する"""
        nodes_set = set()
        for u, v, weight in edges:
            nodes_set.add(u)
            nodes_set.add(v)
            # 双方向のエッジを追加（より長い距離で上書き）
            if v not in self.graph[u] or self.graph[u][v] < weight:
                self.graph[u][v] = weight
            if u not in self.graph[v] or self.graph[v][u] < weight:
                self.graph[v][u] = weight
        
        self.nodes = sorted(list(nodes_set))
        self.node_to_idx = {node: i for i, node in enumerate(self.nodes)}

    def _initialize_dp(self) -> None:
        """DPテーブルを初期化する"""
        INF = float('-inf')
        self.dp = [[INF] * self.n for _ in range(1 << self.n)]
        self.parent = [[None] * self.n for _ in range(1 << self.n)]
        self.start_node = [[-1] * self.n for _ in range(1 << self.n)]

        for i in range(self.n):
            mask = 1 << i
            self.dp[mask][i] = 0.0
            self.start_node[mask][i] = i
            if self.n == 1:
                self.best_mask = mask
                self.best_end_idx = i

    # ----------------------------------------
    # DP実行ロジック
    # ----------------------------------------

    def _update_dp_state(self, new_mask: int, u_idx: int, new_dist: float) -> None:
        """DPテーブルを新しい距離で更新する"""
        # このメソッドは引数が多い元のupdate_dp_tableの一部を担う
        # prev_mask, v_idx, curr_startは呼び出し側で分かっている
        self.dp[new_mask][u_idx] = new_dist
    
    def _try_close_cycle(self, v_idx: int, mask: int) -> None:
        """訪問済み隣接ノードが始点の場合、閉路を試す"""
        visited_count = bin(mask).count('1')
        if visited_count < 3:
            return

        curr_dist = self.dp[mask][v_idx]
        curr_start_idx = self.start_node[mask][v_idx]
        
        v_node = self.nodes[v_idx]
        start_node = self.nodes[curr_start_idx]

        if start_node in self.graph[v_node]:
            cycle_dist = curr_dist + self.graph[v_node][start_node]
            if cycle_dist > self.max_distance:
                self.max_distance = cycle_dist
                self.best_mask = mask
                self.best_end_idx = v_idx
                self.best_is_cycle = True

    def _extend_path(self, v_idx: int, u_idx: int, mask: int) -> None:
        """未訪問の隣接ノードへパスを伸ばす"""
        curr_dist = self.dp[mask][v_idx]
        curr_start_idx = self.start_node[mask][v_idx]
        
        v_node = self.nodes[v_idx]
        u_node = self.nodes[u_idx]
        
        new_dist = curr_dist + self.graph[v_node][u_node]
        new_mask = mask | (1 << u_idx)

        if new_dist > self.dp[new_mask][u_idx]:
            self._update_dp_state(new_mask, u_idx, new_dist)
            self.start_node[new_mask][u_idx] = curr_start_idx
            self.parent[new_mask][u_idx] = (mask, v_idx)
            
            if new_dist > self.max_distance:
                self.max_distance = new_dist
                self.best_mask = new_mask
                self.best_end_idx = u_idx
                self.best_is_cycle = False

    def _process_neighbors(self, v_idx: int, mask: int) -> None:
        """ある状態(mask, v_idx)から遷移可能なすべての隣人を処理する"""
        v_node = self.nodes[v_idx]
        curr_start_idx = self.start_node[mask][v_idx]

        for u_node in self.graph[v_node]:
            u_idx = self.node_to_idx[u_node]
            is_visited = (mask >> u_idx) & 1
            
            if is_visited:
                if u_idx == curr_start_idx:
                    self._try_close_cycle(v_idx, mask)
            else:
                self._extend_path(v_idx, u_idx, mask)

    def solve(self) -> List[int]:
        """最長経路探索を実行し、結果のパスを返す"""
        if self.n == 0:
            return []
        if self.n == 1:
            return [self.nodes[0]]

        for mask in range(1, 1 << self.n):
            for v_idx in range(self.n):
                # 有効な状態 (mask, v_idx) のみ処理
                if self.dp[mask][v_idx] != float('-inf'):
                    self._process_neighbors(v_idx, mask)
        
        return self._reconstruct_path()

    # ----------------------------------------
    # 経路復元
    # ----------------------------------------

    def _reconstruct_path(self) -> List[int]:
        """DPテーブルの親ポインタを辿って最長経路を復元する"""
        if self.max_distance == 0.0 and self.n > 1:
             # 連結成分が複数ある場合、単一ノードを返す
            return [self.nodes[0]] if self.nodes else []

        path = []
        mask, v_idx = self.best_mask, self.best_end_idx

        while mask is not None and v_idx is not None:
            path.append(self.nodes[v_idx])
            parent_info = self.parent[mask][v_idx]
            if parent_info is None:
                break
            mask, v_idx = parent_info
        
        path.reverse()

        if self.best_is_cycle and len(path) > 0:
            path.append(path[0])
            
        return path

# =======================================
# メイン実行ブロック
# =======================================

def main() -> None:
    """メイン関数"""
    try:
        input_lines = sys.stdin.readlines()
        edges = parse_input(input_lines)
        
        if not edges:
            print("Error: No valid input data", file=sys.stderr)
            sys.exit(1)
            
        solver = LongestPathSolver(edges)
        path = solver.solve()
        
        for node in path:
            print(node, end='\r\n')

    except Exception as e:
        print(f"An unexpected error occurred: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: libft_solution/graph_solution.c
#include "libft/libft.h"
#include <math.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_NODES 20
#define MAX_EDGES 200
#define MAX_NODE_ID 1024

typedef struct s_edge
{
	int			u;
	int			v;
	double		weight;
}				t_Edge;

typedef struct s_parent
{
	int			mask;
	int			node_idx;
}				t_Parent;

typedef struct s_solver
{
	double		graph[MAX_NODES][MAX_NODES];
	int			nodes[MAX_NODES];
	int			node_to_idx[MAX_NODE_ID];
	int			num_nodes;
	double		dp[1 << MAX_NODES][MAX_NODES];
	t_Parent	parent[1 << MAX_NODES][MAX_NODES];
	int			start_node[1 << MAX_NODES][MAX_NODES];
	double		max_distance;
	int			best_mask;
	int			best_end_idx;
	bool		best_is_cycle;
}				t_Solver;

void	init_solver(t_Solver *solver)
{
	int	i;
	int	j;

	solver->num_nodes = 0;
	solver->max_distance = 0.0;
	solver->best_mask = 0;
	solver->best_end_idx = 0;
	solver->best_is_cycle = false;
	ft_memset(solver->node_to_idx, -1, sizeof(solver->node_to_idx));
	i = 0;
	while (i < MAX_NODES)
	{
		j = 0;
		while (j < MAX_NODES)
		{
			solver->graph[i][j] = -1.0;
			j++;
		}
		i++;
	}
}

void	add_node(t_Solver *solver, int node_id)
{
	if (node_id >= MAX_NODE_ID)
		return ;
	if (solver->node_to_idx[node_id] == -1)
	{
		if (solver->num_nodes >= MAX_NODES)
			return ;
		solver->node_to_idx[node_id] = solver->num_nodes;
		solver->nodes[solver->num_nodes] = node_id;
		solver->num_nodes++;
	}
}

void	build_graph(t_Solver *solver, t_Edge *edges, int edge_count)
{
	int	i;
	int	u_idx;
	int	v_idx;

	i = 0;
	while (i < edge_count)
	{
		add_node(solver, edges[i].u);
		add_node(solver, edges[i].v);
		i++;
	}
	i = 0;
	while (i < edge_count)
	{
		u_idx = solver->node_to_idx[edges[i].u];
		v_idx = solver->node_to_idx[edges[i].v];
		if (u_idx != -1 && v_idx != -1)
		{
			if (solver->graph[u_idx][v_idx] < edges[i].weight)
			{
				solver->graph[u_idx][v_idx] = edges[i].weight;
				solver->graph[v_idx][u_idx] = edges[i].weight;
			}
		}
		i++;
	}
}

void	init_dp(t_Solver *solver)
{
	int	n;
	int	i;
	int	j;

	n = solver->num_nodes;
	i = 0;
	while (i < (1 << n))
	{
		j = 0;
		while (j < n)
		{
			solver->dp[i][j] = -INFINITY;
			solver->parent[i][j] = (t_Parent){-1, -1};
			solver->start_node[i][j] = -1;
			j++;
		}
		i++;
	}
	i = 0;
	while (i < n)
	{
		solver->dp[1 << i][i] = 0.0;
		solver->start_node[1 << i][i] = i;
		i++;
	}
	if (n == 1)
	{
		solver->best_mask = 1;
		solver->best_end_idx = 0;
	}
}

void	try_close_cycle(t_Solver *solver, int v_idx, int mask)
{
	int		i;
	int		visited_count;
	int		start_idx;
	double	cycle_dist;

	i = 0;
	visited_count = 0;
	while (i < solver->num_nodes)
	{
		if ((mask >> i) & 1)
			visited_count++;
		i++;
	}
	if (visited_count < 3)
		return ;
	start_idx = solver->start_node[mask][v_idx];
	if (solver->graph[v_idx][start_idx] >= 0)
	{
		cycle_dist = solver->dp[mask][v_idx] + solver->graph[v_idx][start_idx];
		if (cycle_dist > solver->max_distance)
		{
			solver->max_distance = cycle_dist;
			solver->best_mask = mask;
			solver->best_end_idx = v_idx;
			solver->best_is_cycle = true;
		}
	}
}

void	extend_path(t_Solver *solver, int v_idx, int u_idx, int mask)
{
	double	new_dist;
	int		new_mask;

	new_dist = solver->dp[mask][v_idx] + solver->graph[v_idx][u_idx];
	new_mask = mask | (1 << u_idx);
	if (new_dist > solver->dp[new_mask][u_idx])
	{
		solver->dp[new_mask][u_idx] = new_dist;
		solver->parent[new_mask][u_idx] = (t_Parent){mask, v_idx};
		solver->start_node[new_mask][u_idx] = solver->start_node[mask][v_idx];
		if (new_dist > solver->max_distance)
		{
			solver->max_distance = new_dist;
			solver->best_mask = new_mask;
			solver->best_end_idx = u_idx;
			solver->best_is_cycle = false;
		}
	}
}

void	process_neighbors(t_Solver *solver, int v_idx, int mask)
{
	int	n;
	int	start_idx;
	int	u_idx;

	n = solver->num_nodes;
	start_idx = solver->start_node[mask][v_idx];
	u_idx = 0;
	while (u_idx < n)
	{
		if (solver->graph[v_idx][u_idx] >= 0)
		{
			if ((mask >> u_idx) & 1)
			{
				if (u_idx == start_idx)
					try_close_cycle(solver, v_idx, mask);
			}
			else
				extend_path(solver, v_idx, u_idx, mask);
		}
		u_idx++;
	}
}

void	solve(t_Solver *solver)
{
	int	n;
	int	mask;
	int	v_idx;

	n = solver->num_nodes;
	if (n == 0 || n == 1)
		return ;
	mask = 1;
	while (mask < (1 << n))
	{
		v_idx = 0;
		while (v_idx < n)
		{
			if (!isinf(solver->dp[mask][v_idx]))
				process_neighbors(solver, v_idx, mask);
			v_idx++;
		}
		mask++;
	}
}

void	reconstruct_and_print_path(t_Solver *solver)
{
	int			path[MAX_NODES];
	int			path_len;
	int			mask;
	int			v_idx;
	t_Parent	p;

	if (solver->num_nodes == 0)
		return ;
	if (solver->max_distance == 0.0 && solver->num_nodes > 1)
	{
		printf("%d\r\n", solver->nodes[0]);
		return ;
	}
	path_len = 0;
	mask = solver->best_mask;
	v_idx = solver->best_end_idx;
	while (v_idx != -1)
	{
		path[path_len++] = solver->nodes[v_idx];
		p = solver->parent[mask][v_idx];
		mask = p.mask;
		v_idx = p.node_idx;
	}
	v_idx = path_len - 1;
	while (v_idx >= 0)
	{
		printf("%d\r\n", path[v_idx]);
		v_idx--;
	}
	if (solver->best_is_cycle && path_len > 0)
		printf("%d\r\n", path[path_len - 1]);
}

void	free_split(char **arr)
{
	int	i;

	if (!arr)
		return ;
	i = 0;
	while (arr[i])
	{
		free(arr[i]);
		i++;
	}
	free(arr);
}

bool	parse_parts(char **parts, t_Edge *edge)
{
	int	count;

	count = 0;
	while (parts[count])
		count++;
	if (count != 3)
		return (false);
	edge->u = ft_atoi(parts[0]);
	edge->v = ft_atoi(parts[1]);
	edge->weight = atof(parts[2]);
	return (true);
}

int	read_input(t_Edge *edges)
{
	char	*line;
	int		count;
	char	**parts;
	int		i;

	count = 0;
	line = get_next_line(0);
	while (line != NULL)
	{
		if (count >= MAX_EDGES)
		{
			free(line);
			break ;
		}
		i = 0;
		while (line[i] && line[i] != '\r' && line[i] != '\n')
			i++;
		line[i] = '\0';
		parts = ft_split(line, ',');
		free(line);
		if (parts)
		{
			if (parse_parts(parts, &edges[count]))
				count++;
			free_split(parts);
		}
		line = get_next_line(0);
	}
	return (count);
}

int	main(int argc, char *argv[])
{
	t_Edge edges[MAX_EDGES];
	int edge_count;
	t_Solver *solver;

	if (argc > 1)
	{
		fprintf(stderr, "Usage: %s < input.txt\n", argv[0]);
		return (1);
	}
	edge_count = read_input(edges);
	if (edge_count == 0)
	{
		ft_putstr_fd("Error: No valid input data\n", 2);
		return (1);
	}
	solver = (t_Solver *)malloc(sizeof(t_Solver));
	if (!solver)
	{
		ft_putstr_fd("Failed to allocate memory for solver\n", 2);
		return (1);
	}
	init_solver(solver);
	build_graph(solver, edges, edge_count);
	init_dp(solver);
	solve(solver);
	reconstruct_and_print_path(solver);
	free(solver);
	return (0);
}